packages/aws-cdk/lib/api/hotswap/lambda-functions.ts

```diff
@@ -0,0 +1,159 @@
+  } else {
+    // verify that the Asset changed - otherwise,
+    // it's a Code property-only change,
+    // but not to an asset change
+    // (for example, going from Code.fromAsset() to Code.fromInline())
+    if (!assetMetadataChanged(change)) {
+      return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+    }
+
+    let functionPhysicalName: string | undefined;
+    try {
+      functionPhysicalName = stringifyPotentialCfnExpression(change.newValue?.Properties?.FunctionName, assetParamsWithEnv);
+    } catch (e) {
+      // It's possible we can't evaluate the function's name -
+      // for example, it can use a Ref to a different resource,
+      // which we wouldn't have in `assetParamsWithEnv`.
+      // That's fine though - ignore any errors,
+      // and treat this case the same way as if the name wasn't provided at all,
+      // which means it will be looked up using the listStackResources() call
+      // by the later phase (which actually does the Lambda function update)
+      functionPhysicalName = undefined;
+    }
+
+    return new LambdaFunctionHotswapOperation({
+      logicalId,
+      physicalName: functionPhysicalName,
+      code: lambdaCodeChange,
+    });
+  }
+}
+
+/**
+ * Returns `true` if the change is not for a AWS::Lambda::Function,
+ * but doesn't prevent short-circuiting
+ * (like a change to CDKMetadata resource),
+ * `false` if the change is to a AWS::Lambda::Function,
+ * but not only to its Code property,
+ * or a LambdaFunctionCode if the change is to a AWS::Lambda::Function,
+ * and only affects its Code property.
+ */
+function isLambdaFunctionCodeOnlyChange(
+  change: cfn_diff.ResourceDifference, assetParamsWithEnv: { [key: string]: string },
+): LambdaFunctionCode | ChangeHotswapImpact {
+  if (!change.newValue) {
+    return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+  }
+  const newResourceType = change.newValue.Type;
+  // Ignore Metadata changes
+  if (newResourceType === 'AWS::CDK::Metadata') {
+    return ChangeHotswapImpact.IRRELEVANT;
+  }
+  // The only other resource change we should see is a Lambda function
+  if (newResourceType !== 'AWS::Lambda::Function') {
+    return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+  }
+  if (change.oldValue?.Type == null) {
+    // this means this is a brand-new Lambda function -
+    // obviously, we can't short-circuit that!
+    return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+  }
+
+  /*
+   * On first glance, we would want to initialize these using the "previous" values (change.oldValue),
+   * in case only one of them changed, like the key, and the Bucket stayed the same.
+   * However, that actually fails for old-style synthesis, which uses CFN Parameters!
+   * Because the names of the Parameters depend on the hash of the Asset,
+   * the Parameters used for the "old" values no longer exist in `assetParams` at this point,
+   * which means we don't have the correct values available to evaluate the CFN expression with.
```

Comment:
I'm not understanding the described logic of "there is a problem/no actually there's not", but if you do need the actual original values, you can get them by describing the stack to get the original params and doing a local CFN eval for the substitution.

Reply:
We don't need the original values, as the comment explains (or, at least tried to).